pkg/client/client.go

```diff
@@ -205,21 +208,104 @@ func NewClientFromKeys(accessKey, secretAccessKey, region string) (Client, error
 	}, nil
 }
 
+// DescribeRegionsData holds output of DescribeRegions API call and the time when it was last updated.
+type DescribeRegionsData struct {
+	err                   error
+	describeRegionsOutput *ec2.DescribeRegionsOutput
+	lastUpdated           time.Time
+}
+
+// cachedDescribeRegions returns full list of regions from AWS.
+// DescribeRegionsOutput is cached to avoid AWS API calls on each reconcile loop.
+func cachedAWSDescribeRegions(awsSession *session.Session, regionsCacheMap map[string]DescribeRegionsData) (*ec2.DescribeRegionsOutput, error) {
+	creds, err := awsSession.Config.Credentials.Get()
+	if err != nil {
+		return nil, err
+	}
+	accessKeyID := creds.AccessKeyID
+	regionsCache, found := regionsCacheMap[accessKeyID]
+	if !found {
+		regionsCache = DescribeRegionsData{}
+	}
+
+	if regionsCache.describeRegionsOutput != nil && regionsCache.err == nil &&
+		time.Since(regionsCache.lastUpdated) < awsRegionsCacheExpirationDuration {
+		klog.Info("Using cached AWS region data")
+		return regionsCache.describeRegionsOutput, regionsCache.err
+	}
+
+	currentRegion := awsSession.Config.Region
+	// Use known endpoint to send our request
+	awsSession.Config.Region = aws.String("us-east-1")
+	describeRegionsOutput, err := ec2.New(awsSession).DescribeRegions(&ec2.DescribeRegionsInput{
+		AllRegions: aws.Bool(true),
+		DryRun:     aws.Bool(false),
+	})
+	// Restore the original region
+	awsSession.Config.Region = currentRegion
+	if err != nil {
+		regionsCache.err = err
+		return nil, err
+	}
+
+	regionsCache.describeRegionsOutput = describeRegionsOutput
```

Comment:
Do we need to handle concurrent access? The machines are reconciled sequentially. Is this just future proofing in case we decide to reconcile multiple MachineSets at once?

Reply:
To a degree it's future proofing, but as the `MaxConcurrentReconciles` is something that could be configured within the controller manager to be more than 1 with ease, it's very easy for this to become a problem in short order